packages/wallets/src/evm/wallet.ts

```diff
@@ -26,48 +68,318 @@ export interface ViemWallet {
+        if (transactionHash === undefined) {
+            throw new Error("Transaction hash not found");
+        }
+        return transactionHash as Hex;
+    }
+
+    private get walletLocator(): string {
+        if (this.apiClient.isServerSide) {
+            return this.address;
+        } else {
+            return "evm-smart-wallet";
+        }
+    }
+
+    private get signerLocator(): string {
+        switch (this.adminSigner.type) {
+            case "evm-passkey":
+                return `evm-passkey:${this.adminSigner.id}`;
+            case "evm-keypair":
+                return `evm-keypair:${this.adminSigner.address}`;
+        }
+        throw new Error("Invalid signer type");
+    }
+
+    private async approveTransaction(transactionId: string, message: Hex) {
+        switch (this.adminSigner.type) {
+            case "evm-passkey": {
+                const { metadata, signature } = await WebAuthnP256.sign({
+                    credentialId: this.adminSigner.id,
+                    challenge: message,
+                });
+
+                await this.apiClient.approveTransaction(this.walletLocator, transactionId, {
+                    approvals: [
+                        {
+                            signer: `evm-passkey:${this.adminSigner.id}`,
+                            // @ts-ignore the generated types are wrong
+                            signature: {
+                                r: `0x${signature.r.toString(16)}`,
+                                s: `0x${signature.s.toString(16)}`,
+                            },
+                            metadata,
+                        },
+                    ],
+                });
+                return;
+            }
+            case "evm-keypair": {
+                const signerAddress = this.adminSigner.address as Address;
+                const signature =
+                    this.adminSigner.signer.type === "viem"
+                        ? await this.adminSigner.signer.account.signMessage!({
+                              message: {
+                                  raw: message,
+                              },
+                          })
+                        : await this.adminSigner.signer.provider.request({
+                              method: "personal_sign",
+                              params: [message, signerAddress],
+                          });
+                await this.apiClient.approveTransaction(this.walletLocator, transactionId, {
+                    approvals: [
+                        {
+                            signature,
+                            signer: `evm-keypair:${signerAddress}`,
+                        },
+                    ],
+                });
+            }
+        }
+    }
+
+    private async approveSignature(signatureId: string, message: Hex) {
+        switch (this.adminSigner.type) {
+            case "evm-passkey": {
+                const { metadata, signature } = await WebAuthnP256.sign({
```

Comment:
could we abstract all the signing logic as it's the same for both txs and signatures?

Reply:
yup https://github.com/Crossmint/crossmint-sdk/pull/970/commits/ff08c0be2000caee5452b6562d1c0e9e6a942c52